### PR TITLE
Rescale covariance in fast MC

### PR DESCRIPTION
**Pull request by mblomqvist**
Adds the option to rescale the covariance matrix inside fast MC for fitter2. Useful for making simple parameter projections. Set the optional scale factor in chi2.ini using e.g. (add more values for multi-fits)

[fast mc]
covscaling = 0.5

The PR also saves the chi2 of every fast MC fit.

### CODE REVIEW

File: py/picca/fitter2/chi2.py [hunk 1]
```diff
@@ -184,12 +186,18 @@ def fastMC(self):
...
+        if hasattr(self,"scalefast_mc"):
+            for d, s in zip(self.data, self.scalefast_mc):
+                d.co = s*d.co
+                d.ico = d.ico/s
+
         for d in self.data:
             d.cho = cholesky(d.co)
 
         self.fast_mc = {}
+        self.fast_mc['chi2'] = []
```
Comment (londumas) [inline py/picca/fitter2/chi2.py]: 'chi2' can be a parameter in fast_mc

File: py/picca/fitter2/chi2.py [hunk 2]
```diff
@@ -31,6 +31,8 @@ def __init__(self,dic_init):
             else:
                 self.seedfast_mc = 0
             self.nfast_mc = dic_init['fast mc']['niterations']
+            if 'covscaling' in dic_init['fast mc']:
+                self.scalefast_mc = dic_init['fast mc']['covscaling']
```
Comment (londumas) [inline py/picca/fitter2/chi2.py]: set to 1 if not given

File: py/picca/fitter2/chi2.py [hunk 3]
```diff
@@ -184,12 +186,18 @@ def fastMC(self):
         if not hasattr(self,"nfast_mc"): return
 
         sp.random.seed(self.seedfast_mc)
-
         nfast_mc = self.nfast_mc
+
+        if hasattr(self,"scalefast_mc"):
```
Comment (londumas) [inline py/picca/fitter2/chi2.py]: always present, but `1` if not given

File: py/picca/fitter2/chi2.py [hunk 4]
```diff
@@ -203,6 +211,7 @@ def fastMC(self):
                 if not p in self.fast_mc:
                     self.fast_mc[p] = []
                 self.fast_mc[p].append([v, best_fit.errors[p]])
+            self.fast_mc['chi2'].append(best_fit.fval)
```
Comment (londumas) [inline py/picca/fitter2/chi2.py]: self.fast_mc[p].append([ sp.append(v,chi2), sp.append(best_fit.errors[p],0.)])

File: py/picca/fitter2/chi2.py [hunk 5]
```diff
@@ -271,12 +280,18 @@ def export(self):
             g = f.create_group("fast mc")
             g.attrs['niterations'] = self.nfast_mc
             g.attrs['seed'] = self.seedfast_mc
+            if hasattr(self, "scalefast_mc"):
```
Comment (londumas) [inline py/picca/fitter2/chi2.py]: always present

File: py/picca/fitter2/chi2.py (hunk 1 above)
Comment (mblomqvist) [reply py/picca/fitter2/chi2.py]: Is it more intuitive that way?

Comment (londumas) [reply py/picca/fitter2/chi2.py]: it is.

File: py/picca/fitter2/chi2.py (hunk 2 above)
Comment (mblomqvist) [reply py/picca/fitter2/chi2.py]: I considered it but found it unnecessary to define and store it if it's unity (we don't store an empty list of priors when there are none). I can change it though if you want.

Comment (londumas) [reply py/picca/fitter2/chi2.py]: it is simply that with setting it to 1 by default you don't have to use if hasattr(self,"scalefast_mc").
As you want.

File: py/picca/fitter2/chi2.py (hunk 4 above)
Comment (mblomqvist) [reply py/picca/fitter2/chi2.py]: I don't see how your suggestion fits in here

Comment (londumas) [reply py/picca/fitter2/chi2.py]: it doesn't anymore.

Comment (mblomqvist) [reply py/picca/fitter2/chi2.py]: How do you mean?

Comment (londumas) [reply py/picca/fitter2/chi2.py]: forget about my comment

**Comment by mblomqvist** [comment]
Update to always define and save the covscaling. Codacy doesn't like 'assert', as we know.

**Comment by londumas** [comment]
@mblomqvist, can you try what codacy is recommending in replacement of the assert?

**Comment by ngbusca** [comment]
@mblomqvist @londumas I'll go through this PR today. Regarding the use of `assert`, it's standard python and we shouldn't be avoiding it. If coadacy complains, so be it...

**Comment by londumas** [comment]
@ngbusca, I agree. However if we can do another way, why not.

**Comment by mblomqvist** [comment]
@ngbusca ready to be merged?

**Pull request merged**